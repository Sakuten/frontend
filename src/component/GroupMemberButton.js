import React from 'react'
import { observable, action } from 'mobx'
import { observer } from 'mobx-react'

import {extractId} from '../util/extractId'

import QRReader from '../component/QRReader'

@observer class GroupMemberButton extends React.Component {
  @observable isScanning = false

  render () {
    return (
      <div>
        {
          this.isScanning
            ? <div>
              <QRReader
                onError={this.onError}
                onScan={this.onScan}
              />
              <button className='button' data-test='groupmemberbutton-cancel' onClick={this.onCancel} >キャンセル</button>
            </div>
            : <button className='button' data-test='groupmemberbutton-button' onClick={this.onClick} >{this.props.children}</button>
        }
      </div>
    )
  }

  @action.bound
  onError (error) {
    if (typeof this.props.onError === 'function') {
      this.props.onError(error)
    }
  }

  @action.bound
  onScan (scanUri) {
    if (scanUri) {
      const secretId = extractId(scanUri)
      if (!secretId) {
        this.store.error.addError('Invalid QR Code')
        return
      }

      this.props.onAdd(secretId)
      this.isScanning = false
    }
  }

  @action.bound
  onClick () {
    this.isScanning = true
  }

  @action.bound
  onCancel () {
    this.isScanning = false
  }
}

export default GroupMemberButton
