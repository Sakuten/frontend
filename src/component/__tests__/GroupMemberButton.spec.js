import React from 'react'
import { shallow } from 'enzyme'
import GroupMemberButton from '../GroupMemberButton'

const setup = propOverrides => {
  const props = Object.assign({
    onAdd: jest.fn()
  }, propOverrides)

  const wrapper = shallow(<GroupMemberButton {...props} />)

  return {
    props,
    wrapper,
    button: wrapper.find('[data-test="groupmemberbutton-button"]')
  }
}

describe('components', () => {
  describe('GroupMemberButton', () => {
    it('render', () => {
      const { wrapper } = setup()
      expect(wrapper).toMatchSnapshot()
    })

    it('renders button', () => {
      const { button } = setup()
      expect(button.exists()).toBe(true)
    })

    it('changes view when button is clicked', () => {
      const { wrapper, button } = setup()
      button.simulate('click')
      // refactoring required: reuse `button` to see the changes
      expect(wrapper.find('[data-test="groupmemberbutton-button"]').exists()).toBe(false)
    })
  })
})
