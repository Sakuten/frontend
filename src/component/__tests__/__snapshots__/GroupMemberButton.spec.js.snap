// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components GroupMemberButton render 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <GroupMemberButton
    onAdd={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <button
        data-test="groupmemberbutton-button"
        onClick={[Function]}
      />,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": undefined,
        "data-test": "groupmemberbutton-button",
        "onClick": [Function],
      },
      "ref": null,
      "rendered": null,
      "type": "button",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <button
          data-test="groupmemberbutton-button"
          onClick={[Function]}
        />,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": undefined,
          "data-test": "groupmemberbutton-button",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": "button",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
